ts/editor/destroyable.ts

```diff
@@ -0,0 +1,35 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+export interface Destroyable {
+    destroy(): void;
+}
+
+export function clearableArray<T>(): (T & Destroyable)[] {
```

Comment:
A docstring on the function would be nice. 

Is there any risk of the cached destroy method pointing to the wrong place if something is manually inserted into or removed from the list?

Reply:
The index is only created inside the destroy function, so I'd say no.
Something that could be awkward is if somebody tried to check for the existence of a field:

```
if (field[index]) { /* do something */ }
// this will always be correct, instead they should do this:
if (index in field) { /* do something */ }
```

Reply:
Hmm, I'm not fully understanding the purpose of this function. Is my understanding correct that it tries to catch numeric references (eg list[5]), and if they are false-ish, it returns an empty object with a destroy method that will remove the entry from the list and shift other entries up? If so, that seems a bit complicated and magical - is it not at all practical to do this a different way that is more explicit?

Reply:
I will give another look over the coming days, and see if I can refactor in. If I cannot, I'll explain in more detail why not.